beep/generate_protocol.py

```diff
@@ -988,9 +989,70 @@ def convert_velocity_to_power_waveform(waveform_file, velocity_units):
 
     power = -power * df.v  # positive power = charge
 
-    return pd.DataFrame({'time(s)': df.t,
-                         'power_unscaled(W)': power,
-                         'power_scaled': power / max(abs(power))})
+    return pd.DataFrame({'time': df.t,
+                         'power': power})
+
+
+def generate_maccor_waveform_file(df, file_prefix, file_directory, mwf_config=None):
+    """
+    Helper function that takes in a variable power waveform and outputs a maccor waveform file (.MWF), which is read by
+    the cycler when the procedure file has a "fast-wave" step.
+    Relevant parameters to generate the .mwf files governing the input mode, charge/discharge limits, end conditions
+    and scaling are stored in /conversion_schemas/maccor_waveform_conversion.yaml
+
+    Args:
+        df (pd.DataFrame): power waveform containing two columns "time" and "power", in sec and W respectively.
+        file_prefix (str): prefix for the filename (extension is .MWF by default)
+        file_directory (str): folder to store the mwf file
+        mwf_config (dict): dictionary of instrument control parameters for generating the waveform
+
+    Returns:
+         path to the maccor waveform file generated
+    """
+
+    if mwf_config is None:
+        mwf_config = MACCOR_WAVEFORM_CONFIG
+
+    df['power'] = df['power'] / max(abs(df['power'])) * mwf_config['value_scale']
+
+    df['step_counter'] = df['power'].diff().fillna(0).ne(0).cumsum()
+    df = df.groupby('step_counter').agg({'time': 'count',
+                                         'power': 'first'})
+
+    df.rename(columns={'time': 'duration'}, inplace=True)
+
+    df['control_mode'] = mwf_config['control_mode']
+
+    if mwf_config['control_mode'] == 'P':
+        df.loc[df['power'] == 0, 'control_mode'] = 'I'
+
+    df['value'] = np.round(abs(df['power']), 5)
+
+    mask = df['power'] <= 0
+    df = df.assign(**{'state': 'C',
+                      'limit_mode': mwf_config['charge_limit_mode'],
+                      'limit_value': mwf_config['charge_limit_value'],
+                      'end_mode': mwf_config['charge_end_mode'],
+                      'operation':  mwf_config['charge_end_operation'],
+                      'end_mode_value': mwf_config['charge_end_mode_value']
+                      })
+
+    df.loc[mask, ['state', 'limit_mode', 'limit_value', 'end_mode', 'operation', 'end_mode_value']] =\
+        ['D', mwf_config['discharge_limit_mode'], mwf_config['discharge_limit_value'],
```

Comment:
Charge and discharge voltage limits might need to be changed from cell to cell, my guess is that we will pass them into the function as a portion of the parameters in the protocol. We can do that in a later PR?

Reply:
Or actually, since you are already using the config, we can just send a different config when call this function.

Reply:
Yeah, there should be a method in the `ProcedureFile` class that calls this helper, and we will modify the default configuration as appropriate before doing so.